Move editor preference option to Site Settings screen

daniloercoli opened the pull request:
Move the "default editor for new posts" toggle from App Settings screen to Site Settings screen.

Details here: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1233

**Test 1** - Account already migrated to version 12.9 and that has seen the auto-enable dialog
(the app has the global option set to TRUE)
- Open app settings and see there is no editor option
- Go to site settings. The default editor preference is there and it should be `TRUE`
- Change it to false
- Re-open settings, it should show the correct value
- Tap to create a new post, the right editor (Aztec) should be shown

**Test 2** - Account already migrated to version 12.9 that has not yet seen the auto-enable dialog
_(so basically an account that didn't open a GB based post in v12.9)_
- Open app settings and see there is no editor option
- Go to site settings. The default editor preference is there and it should be `false`
- Tap to create a new post, the right editor (Aztec) should be shown
- Open a blocks based post 
- The dialog that informs of the auto-enable should appear 
- Re-open settings, it should show the correct value (`true`)
- Tap to create a new post, the right editor (GB) should be shown.

**Test 3** - Account already migrated to version 12.9 that has set the editor option to `FALSE` after the auto-opt in dialog.
_(so basically an account that opened a GB based post in v12.9, but turned it OFF later)_
- Open app settings and see there is no editor option
- Go to site settings. The default editor preference is there and it should be `false`
- Tap to create a new post, the right editor (Aztec) should be shown
- Open a blocks based post 
- NO dialog that informs of the auto-enable should appear 
- Re-open settings, it should show the correct value (`false`)
- Tap to create a new post, the right editor (Aztec) should be shown.

**Test 4** - NEW account
- Open app settings and see there is no editor option
- Go to site settings. The default editor preference is there and it should be `true`
- Tap to create a new post, the right editor (GB) should be shown
- Go to site settings and change it to false
- Re-open settings, it should show the correct value
- Tap to create a new post, the right editor (Aztec) should be shown.

**Test 5** - Self hosted sites - Upgrading from 12.9 where the block editor was enabled globally
- Open app settings and see there is no editor option
- Go to site settings. The default editor preference is there and it should be `true`
- Tap to create a new post, the right editor (GB) should be shown
- Go to site settings and change it to false
- Re-open settings, it should show the correct value
- Tap to create a new post, the right editor (Aztec) should be shown

**Test 6** - Self hosted sites - new install
- Open app settings and see there is no editor option
- Go to site settings. The default editor preference is there and it should be `false`
- Tap to create a new post, the right editor (Aztec) should be shown
- Go to site settings and change it to true
- Re-open settings, it should show the correct value
- Tap to create a new post, the right editor (GB) should be shown

Update release notes:

- [ x ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

hypest commented on WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java:
```diff
@@ -6,14 +6,35 @@
 public class SiteUtils {
+    public static final int MIN_WPCOM_USER_ID_TO_DEFAULT_GB = 149287441;
+
+    public static boolean isBlockEditorDefaultForNewPost(SiteModel site, AccountStore accountStore) {
```
It seems that the `accountStore` can be `null` and is even marked as nullable [here](https://github.com/wordpress-mobile/WordPress-Android/blob/673115d1ece0620acd0e181721584d3f72a3742f/WordPress/src/main/java/org/wordpress/android/ui/accounts/HelpActivity.kt#L171) and [here](https://github.com/wordpress-mobile/WordPress-Android/blob/5573e47b1c97da0834c66b6772923e7ffc015a17/WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java#L629). My impression is that it shouldn't be nullable at all since the Stores are supposed to be available.

hypest replied:
Apparently fixed with d2f8442d2c80ecff47d95fb0b8ba4fe9caeaef68.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1254,7 +1256,8 @@ public boolean onPrepareOptionsMenu(Menu menu) {
             //  "Classic Block" support yet so, don't offer a switch to it if content doesn't have blocks. If the post
             //  is empty but the user hasn't enabled "Use Gutenberg for new posts" App setting, don't offer the switch.
-            switchToGutenbergMenuItem.setVisible(hasBlocks || (AppPrefs.isGutenbergDefaultForNewPosts() && isEmpty));
+            switchToGutenbergMenuItem.setVisible(
```
Let's also update the comment in the above line (line number 1258) to mention "Site setting" instead of "App setting".

hypest replied:
Fixed with 1a06e9111892b2bc155af716215b342b0bb0a57f.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1541,8 +1544,10 @@ private void showGutenbergInformativeDialog() {
     private void setGutenbergEnabledIfNeeded() {
         if (AppPrefs.isGutenbergAutoEnabledForTheNewPosts()
```
Kinda a nit and I know it's not part of the changeset in this PR but, I find the `isGutenbergAutoEnabledForTheNewPosts` name a bit confusing, especially because of the past tense in `Enabled`. Reading it I thought it would be true if the Gutenberg has been enabled automatically in the past but that's not its implementation. Can we change it to `shouldAutoEnableGutenbergForTheNewPosts`?

hypest replied:
Renamed with 0cf405d4ec264499095cbf55ccee3a61837e82c8.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostUtils.java:
```diff
@@ -153,11 +154,6 @@ public static void trackSavePostAnalytics(PostModel post, SiteModel site) {
                     AnalyticsUtils.trackWithSiteDetails(AnalyticsTracker.Stat.EDITOR_UPDATED_POST, site, properties);
                 } else {
                     properties.put("word_count", AnalyticsUtils.getWordCount(post.getContent()));
-                    properties.put("editor_source",
```
Hmm, did we remove the `editor_source` property on purpose or do we need to migrate how to compute it?

daniloercoli replied:
yes, I left `editor_source` out for a late pass over it. 
It's quite hard to make it working fine on 100% of cases, but I think i made it working for most of  those  in https://github.com/wordpress-mobile/WordPress-Android/pull/10254/commits/af0d48f6c9b8b58530e3748b7f2ba2faabb535d1

The main problem with this part of the analytics code is that, under some circumstances, we need to access the account store (when the site editor preference is still empty for example) and by using the `user_id` decide if the block editor is the default editor or not. It's a very small % of cases I think, since the migration should have already happened at this point, but it could happen...
On the commit above i've passed the AccountStore to this analytics utils where possible, keeping it null on some cases where introducing it would require lot of refactoring.

hypest replied:
> keeping it null on some cases where introducing it would require lot of refactoring

Aha, had a look and [this one](https://github.com/wordpress-mobile/WordPress-Android/commit/af0d48f6c9b8b58530e3748b7f2ba2faabb535d1#diff-6f326538ef251c6604082c8eb857f51bR69) is quite trivial (we just need to pass the AccountStore in the constructor and there's only one call site to update).

[The other case](https://github.com/wordpress-mobile/WordPress-Android/commit/af0d48f6c9b8b58530e3748b7f2ba2faabb535d1#diff-07a4b2d77083cfd60941cda4d2aa9a00R265), does indeed require some refactoring.

daniloercoli replied:
Ok, so let's mark this resolved since we've settled in a Slack convo to not consider the user_id during this migration, resulting in clearing the AccountStore parameter and fix this convo as well.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java:
```diff
@@ -732,6 +734,11 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
             mSiteSettings.setTimezone(newValue.toString());
+        } else if (preference == mGutenbergDefaultForNewPosts) {
+            mDispatcher.dispatch(SiteActionBuilder.newDesignateMobileEditorAction(
+                    new DesignateMobileEditorPayload(mSite, ((Boolean) newValue) ? "gutenberg" : "aztec")));
```
WDYT about declaring the editor payload literals ("gutenberg" and "aztec") somewhere as string constants? It seems they are used in a few different places throughout the codebase.

hypest replied:
Fixed with 77f01f06578772c079abee27e94a8076292d338c.

hypest commented on WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java:
```diff
@@ -105,7 +104,11 @@ public static void refreshMetadata(AccountStore accountStore, SiteStore siteStor
         metadata.setUsername(accountStore.getAccount().getUserName());
         metadata.setEmail(accountStore.getAccount().getEmail());
-        metadata.setGutenbergEnabled(AppPrefs.isGutenbergDefaultForNewPosts());
+        // Use the first site in the list. Should be the default
```
I know this will be accurate for users that have only one site but, feels weird to just pick the first site from the list to set the `gutenberg_enabled` Tracks metadata field. The natural thing to do would be to just pick up the currently selected site, right?

Perhaps we can use [AppPrefs.getSelectedSite()](https://github.com/wordpress-mobile/WordPress-Android/blob/18a236470a673a6f19b7dea8feb4b79330ef80d2/WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java#L550) and `mSiteStore.getSiteByLocalId()` to get the SiteModel and from there the flag?

Just to be sure we're in sync here, where will the `gutenberg_enabled` Tracks metadata flag come from in WPiOS @etoledom ?

daniloercoli replied:
Right, nice call. Will change the code and get the current selected site. cc @etoledom

hypest replied:
Seems fixed with b66a9c52af8a46b1ba23a58960c01774ff1db4d2.

daniloercoli replied:
yes, it was fixed there, left it unsolved to make sure @etoledom will see the changes.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1541,8 +1544,10 @@ private void showGutenbergInformativeDialog() {
             && !mIsNewPost
-            && !AppPrefs.isGutenbergDefaultForNewPosts()) {
+            && !SiteUtils.isBlockEditorDefaultForNewPost(mSite, mAccountStore)) {
             AppPrefs.setGutenbergDefaultForNewPosts(true);
```
Do we actually need the app-wide Gutenberg pref anymore? Can we just remove it altogether?

daniloercoli replied:
No, we don't need it anymore. Removed.

hypest replied:
Removed with 5680f4da63f96907a02755aacf6b4893d190a998.

hypest commented on WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java:
```diff
@@ -127,7 +130,8 @@ public static void refreshMetadataNewUser(String username, String email) {
         metadata.setEmail(email);
-        metadata.setGutenbergEnabled(AppPrefs.isGutenbergDefaultForNewPosts());
+        // GB is enabled for new users
+        metadata.setGutenbergEnabled(true);
```
Shouldn't this be reflected somehow in the implementation of [SiteUtils.isBlockEditorDefaultForNewPost](https://github.com/wordpress-mobile/WordPress-Android/blob/3346b32c858c26f5200aa2a87895c5f1f259e8aa/WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java#L21) too?

daniloercoli replied:
This code is used when a new account is created. It's "kind" of safe to set it to `true` here, since we've enabled the block editor on new users.
We can although change and use the `SiteUtils.isBlockEditorDefaultForNewPost ` method call, but change may not be super trivial since a call comes from the LoginLibrary. 
Need to check this more deeper though.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java:
```diff
@@ -280,16 +280,6 @@ private void prepareUploadAnalytics(String postContent) {
                 sCurrentUploadingPostAnalyticsProperties = new HashMap<>();
                 sCurrentUploadingPostAnalyticsProperties
                         .put("word_count", AnalyticsUtils.getWordCount(mPost.getContent()));
-                sCurrentUploadingPostAnalyticsProperties.put("editor_source",
```
Similar to a comment earlier, do we intentionally remove the `editor_source` property?

hypest commented on build.gradle:
```diff
@@ -106,5 +106,5 @@ buildScan {
 ext {
     daggerVersion = '2.22.1'
-    fluxCVersion = '1.0.0'
+    fluxCVersion = '8fd938bfaa849fc2ecdc30cd72037ce2f1cff5d1'
```
Let's remember to publish a new tagged FluxC version and point to it before merging this PR.

daniloercoli commented:
Editorial review - The following are the strings changed in this PR: 

Added to the README section
- `* Block editor: Added the ability to enable/disable the block editor on a per site basis.`

Changed strings in the app:
-  `This post uses the block editor, which is the default editor for new posts. To enable the classic editor, go to Me > App Settings`  ---> `This post uses the block editor, which is the default editor for new posts. To enable the classic editor, go to Site Settings.`

- `This page uses the block editor, which is the default editor for new pages. To enable the classic editor, go to Me > App Settings` ---> `This page uses the block editor, which is the default editor for new pages. To enable the classic editor, go to Site Settings.`

benhuberman commented:
Thanks for the ping, @daniloercoli! 

For the README bit, I'd suggest a tweak to center the user's perspective (and a couple of smaller style fixes):
> Block editor: You can now enable and disable the block editor on a per-site basis.

Somewhat similarly for the new post/page messages, how about:

> You're currently using the block editor to create new [posts/pages]. Want to enable the classic editor? Head to Site Settings.

I don't know that explaining that the block editor is the default for new posts/pages is crucial to mention, but if you think it is, I'm happy to propose an alternative that includes that detail.

hypest commented:
> I don't know that explaining that the block editor is the default for new posts/pages is crucial to mention, but if you think it is, I'm happy to propose an alternative that includes that detail.

Thanks for the proposals @benhuberman ! In this case, I think we need to reinstate a form of the "default" mention. Here's what happens:
1. The block editor offers switching to classic and back to block editor per post anyway, available in the 3-dots menu inside the editor. The production version of this one-time popup mentions that.
2. In this new version, we introduce an auto opt-in, where the first time a block post gets opened we will enable using the block editor _by default_ for new posts. The user can change that setting via the Site Settings in the "My Site" screen. We modify the one-time popup to mention that default change, and the popup replaces the previous one.

So, yes please, let's have a version of the copy that's about the default. Does the following one look good to you? `You're currently using the block editor to create new [posts/pages]. Want to default to the classic editor? Head to Site Settings.`

benhuberman commented:
Would this tweak work for your purposes here, @hypest? It doesn't say "default," but emphasizes that we're talking about a "sticky" change, not just something for one session:

> You're currently using the block editor to create new [posts/pages]. Want to use the classic editor going forward? Head to Site Settings.

If you absolutely want to include "default" in the language, I'd go with:
> You're currently using the block editor to create new [posts/pages]. Want to use the classic editor by default? Head to Site Settings.

A question that just came to me: when looking at the Site Settings screen, is it obvious to the user what they should do? If not, it might make sense to add a secondary instruction, like "Head to **Site Settings > Default editor for new posts.** (This also makes the sentence quite a bit longer, so I'd probably only add if we're really concerned about users not knowing what to do next after going to Site Settings).

hypest commented:
Thanks for the alternatives Ben! Let's go with the `Want to use the classic editor by default?` version since the `going forward` one feels too "terminal" and long term.

@daniloercoli , feel free to do the change of the copy, thanks! cc @etoledom

hypest commented on RELEASE-NOTES.txt:
```diff
@@ -1,5 +1,6 @@
 13.0
 -----
+* Block editor: Added the ability to enable/disable the block editor on a per site basis.
```
How about we also add `Block editor: Default to the block editor unless user opted out in v12.9`?

daniloercoli replied:
Cool! Added in 4ba301608db9cf86e81fab13191903f3684ea0d5.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -658,9 +658,9 @@ public static boolean isGutenbergDefaultForNewPosts() {
     }
 
-    public static void setGutenbergDefaultForNewPosts(boolean defaultForNewPosts) {
-        AnalyticsTracker.track(defaultForNewPosts ? Stat.EDITOR_GUTENBERG_ENABLED : Stat.EDITOR_GUTENBERG_DISABLED);
```
With this Track firing removal, it seems that we never trigger the `EDITOR_GUTENBERG_ENABLED` track. Perhaps we need to move it to the site setting switching?

hypest replied:
cc @etoledom so to be in alignment with the WPiOS side.

daniloercoli replied:
Oh right! 🤦‍♂ Re-added in ee57eb45afeff9daa98df47291e752ab6d3c65ff

Also tracking with site details now.

hypest replied:
Looping back, added a related comment here https://github.com/wordpress-mobile/WordPress-Android/pull/10254/commits/ee57eb45afeff9daa98df47291e752ab6d3c65ff#r307692184.

hypest commented:
Hey @daniloercoli , apart from the [copy update](https://github.com/wordpress-mobile/WordPress-Android/pull/10254#issuecomment-515334567), I think this PR is almost ready to merge. Can you create a FluxC release and point to it so we don't use a hash? Thanks!

daniloercoli commented:
> Can you create a FluxC release and point to it so we don't use a hash? Thanks!

Addressed in ac008196b38817b3307c260af448cdf259dc1495

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsInterface.java:
```diff
@@ -1145,6 +1147,10 @@ public void onSiteEditorsChanged(OnSiteEditorsChanged event) {
                         new DesignateMobileEditorPayload(event.site, defaultEditor)));
+                AnalyticsUtils.trackWithSiteDetails(
+                        AppPrefs.isGutenbergDefaultForNewPosts() ? Stat.EDITOR_GUTENBERG_ENABLED
+                                : Stat.EDITOR_GUTENBERG_DISABLED,
```
Note that if we indeed emit the `_DISABLED` event here, we will probably see a spike from the users that have upgraded to app version 13.0. Those users won't actually have actively disabled the block editor but rather, it's just going to be the app itself broadcasting the fact that the classic editor gets saved to the backend.

I'd prefer if didn't emit this to avoid confusion when reading the tracks but, it's only going to be a "one time" event so probably not a real issue. WDYT?

daniloercoli replied:
Totally agree with you. https://github.com/wordpress-mobile/WordPress-Android/pull/10254/commits/e8f1f2844fa25b0121a2b4752b17df20990c2c04 does fix the problem.

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsInterface.java:
```diff
@@ -30,6 +33,7 @@
 import org.wordpress.android.util.LocaleManager;
 import org.wordpress.android.util.SiteUtils;
 import org.wordpress.android.util.StringUtils;
+import org.wordpress.android.util.analytics.AnalyticsUtils;
```
Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/10254/commits/2183b7f5b3a006c2e14dcdd5a1c009fbfe0c0f14

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -646,12 +648,44 @@ public static boolean isGutenbergEditorEnabled() {
+        if (urls != null) {
+            newUrls.addAll(urls);
+        }
+
```
It seems we don't add the provided `siteURL` though, is that on purpose? 🤔

daniloercoli replied:
Fixed in 6e4623b95dd1e9648e9ac9e7deeb66f4c0c33c80

hypest replied:
Thanks for the fix Danilo!

But now, I'm actually wondering why do we even need the `setGutenbergAutoEnabledForTheNewPosts()` function which keeps flags locally? If no reason at this stage, WDYT about replacing it with a check of the Site Setting for 'emptiness'?

daniloercoli replied:
Right, there is no need for it anymore.
Addressed in 30cd235fe5b44e1f9939f7d684aaeb967fac0f39

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerActivity.java:
```diff
@@ -221,12 +221,13 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
                 if (data != null) {
                     int newSiteLocalID = data.getIntExtra(SitePickerActivity.KEY_LOCAL_ID, -1);
-                    SiteUtils.enableBlockEditorForNewUsers(mDispatcher, mAccountStore, mSiteStore, newSiteLocalID);
+                    SiteUtils.enableBlockEditor(mDispatcher, mSiteStore, newSiteLocalID);
```
WDYT about moving the call closer to the site creation process, right after the site creation success here https://github.com/wordpress-mobile/WordPress-Android/blob/47b9d8091ec6625e1dac98cf658d7f667764113d/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/services/SiteCreationServiceManager.kt#L88?

The process can happen on the background as well and the UI won't get updated until the app returns to the foreground so, better to move the call in the background too?

daniloercoli replied:
Fixed in 6e4623b95dd1e9648e9ac9e7deeb66f4c0c33c80

daniloercoli replied:
Yes, i tried to move it there, but the SiteStore didn't have the site locally yet.
I guess we can go with the current solution and fix it along the way next week. Ideally it should be called right after the site is created in the server.

hypest replied:
> but the SiteStore didn't have the site locally yet

Arghh, right. Bummer.

hypest commented on WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java:
```diff
@@ -21,6 +21,48 @@
+    public static void migrateAppWideMobileEditorPreferenceToRemote(final Dispatcher dispatcher,
+                                                                    final SiteStore siteStore) {
+        if (!AppPrefs.isDefaultAppWideEditorPreferenceSet()) {
+            return;
```
As per the "OPTED OUT" case mentioned in the function comment, we need to check if the user *wasn't* offered auto-enable before to just return, right?

daniloercoli replied:
Right, if the user didn't open a block based post in a previous version of the app, the preference is not there in the app. We will show the opt-it popup as soon as the user open a block based post.

hypest commented:
Tried:
- Create a new WP.com site on mobile and check Gutenberg is default in Site Settings: ✅
- Create a new WP.com site on mobile and start a new post, ensure it's Gutenberg: ✅
- Clean app, clean backend, start new post → should be Aztec ✅
- Clean app, "Aztec" in backend, start new post → should be Aztec: ✅
- Clean app, "Gutenberg" in backend, start new post → should be Gutenberg: ✅
- Clean  app, clean backend, open Gutenberg post → auto-enables Gutenberg: ✅
- Clean app, clean backend, open Gutenberg post (auto-enables GB), new post in other site → should be Aztec: ✅
- Clean app, clean backend, open Gutenberg post (auto-enables GB), open GB post in other site → auto-enables Gutenberg: ✅
- Toggling the setting in a site doesn't toggle it in others: ✅
- Migrate from v12.9 with GB ON, start new post → should be Gutenberg in all sites:
- Migrate from v12.9 with GB OFF and been auto-enabled, start new post → should be Aztec in all sites:
- Migrate from v12.9 with GB OFF and never auto-enabled, start new post → should be Aztec in all sites:
- Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post → auto-enable Gutenberg on the site:
- Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (will auto-enable GB), change to other site and open GB post → auto-enable Gutenberg on the site:

daniloercoli commented:
> Migrate from v12.9 with GB ON, start new post → should be Gutenberg in all sites: ❌ seems to be Aztec on current site, Gutenberg after changing to other site.

Not sure  what is happening here.  For a short time the `mSiteStore` object in `WPMainActivity` is still returning the old version of the SiteModel for the current site. Even if you reload the site info from the DB after the "setDefaultEditor" network call returned with success for the current site, `SiteModel site = mSiteStore.getSiteByLocalId(getSelectedSite().getId());` the  `mSiteStore` does return an old version of the model.
If you switch tab, ie go to Reader, then go back to My site, it then has the updated version, and the app starts GB correctly.

hypest commented:
> For a short time the mSiteStore object in WPMainActivity is still returning the old version of the SiteModel for the current site. Even if you reload the site info from the DB after the "setDefaultEditor" network call returned with success for the current site, SiteModel site = mSiteStore.getSiteByLocalId(getSelectedSite().getId()); the mSiteStore does return an old version of the model.
> If you switch tab, ie go to Reader, then go back to My site, it then has the updated version, and the app starts GB correctly.

Hmm, not sure, maybe @planarvoid or have a clue of what might be happening here and the model object is not updating in time? Does that ring any bells perhaps?

daniloercoli commented:
> > For a short time the mSiteStore object in WPMainActivity is still returning the old version of the SiteModel for the current site. Even if you reload the site info from the DB after the "setDefaultEditor" network call returned with success for the current site, SiteModel site = mSiteStore.getSiteByLocalId(getSelectedSite().getId()); the mSiteStore does return an old version of the model.
> > If you switch tab, ie go to Reader, then go back to My site, it then has the updated version, and the app starts GB correctly.
> 
> Hmm, not sure, maybe @planarvoid or have a clue of what might be happening here and the model object is not updating in time? Does that ring any bells perhaps?

@hypest I've investigated the issue a bit more and seems that `SiteModel` returned by SiteStore doesn't always contain the correct value for `mobile_editor`. Even if the property is correctly stored on the backend. I guess there is a problem on FluxC somewhere, since the `onSiteChanged` event in WPMainActivity here: https://github.com/wordpress-mobile/WordPress-Android/blob/gb/move-editor-preference-toggle-to-site-settings/WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java#L1136 could receive a model with empty value set for editor. 
This is not happening all the time though, and once the value `Gutenberg` is received on the client app it always stays the same. FluxC will continue emitting `onSiteChanged` event with the correct value. 

Anyway, I think we can investigate this better during the next week, and patch the RC.

hypest commented:
I had a look at the code and what the app is trying to do and I think there's maybe a race condition. The app will try to fetch the site editor setting ([see here](https://github.com/wordpress-mobile/WordPress-Android/blob/86416bbe050610ee59e0a5e5ffb52e710f4dff39/WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsInterface.java#L952)) while in the same time try to migrate the previous App setting ([see here](https://github.com/wordpress-mobile/WordPress-Android/blob/86416bbe050610ee59e0a5e5ffb52e710f4dff39/WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java#L50)). The two network calls are in flight at the same time and their response will get saved to the DB. The "designate editor" call seems to finish first and the correct value is written. But then, the other call arrives, and overwrites the value. If you have cleared the backend then that's what stays (`null`).

I added some Logs to see the sequence of events and here's what I saw:
```
2019-07-29 00:14:26.069  Launching designateMobileEditor
2019-07-29 00:14:26.155 Launching fetchSiteEditors
2019-07-29 00:14:27.273 got designate reply: gutenberg
2019-07-29 00:14:27.280 got fetchEditors reply: 
2019-07-29 00:14:27.295 saving to DB: gutenberg
2019-07-29 00:14:27.297 onSiteEditorsChanged https://stefanosuser.wordpress.com/wp-admin/, gutenberg
2019-07-29 00:14:27.298 saving to DB: 
2019-07-29 00:14:27.298 onSiteEditorsChanged https://stefanosuser.wordpress.com/wp-admin/, 
```

hypest commented:
> ⚠️ Create a new WP.com site on mobile and start a new post → should be Gutenberg and show the info popup. Result: it's Gutenberg but no popup.

Added a detail to the testcase which makes the test "yellow": we need to show the info popup when starting a new Gutenberg post on a new site.

daniloercoli commented:
> > ⚠️ Create a new WP.com site on mobile and start a new post → should be Gutenberg and show the info popup. Result: it's Gutenberg but no popup.
> 
> Added a detail to the testcase which makes the test "yellow": we need to show the info popup when starting a new Gutenberg post on a new site.

Fixed in ff934cdd2f30ddf62ee2fcf8efe93bec5970362a

daniloercoli commented:
> @daniloercoli , I had a look at the code and what the app is trying to do and I think there maybe is a race condition. The app will try to fetch the site editor setting ([see here](https://github.com/wordpress-mobile/WordPress-Android/blob/86416bbe050610ee59e0a5e5ffb52e710f4dff39/WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsInterface.java#L952)) while in the same time try to migrate the previous App setting ([see here](https://github.com/wordpress-mobile/WordPress-Android/blob/86416bbe050610ee59e0a5e5ffb52e710f4dff39/WordPress/src/main/java/org/wordpress/android/util/SiteUtils.java#L50)). The two network calls are in flight at the same time and their response will get saved to the DB. The "designate editor" call seems to finish first and the correct value is written. But then, the other call arrives, and overwrites the value. If you have cleared the backend then that's what stays (`null`).
> 
> I added some Logs to see the sequence of events and here's what I saw: (the network calls and DB saves favor the fetchSiteEditors path and the "un-migrated" backend setting is saved last)

These are the steps to replicate the (already [fixed](https://github.com/wordpress-mobile/WordPress-Android/pull/10254/commits/74a3b822b48f36cb61875915380608661a6b72a5)) problem 

1. install v12.9 
2. switch GB on in App Settings
3. stay on the Me tab
4. Run the latest branch via AS
5. Tap on the big "new post" button at the bottom middle
6. GB comes up

hypest commented:
Trying the latest (8b2de41)

## WP.com/Jetpack sites
⏳ Create a new WP.com site on mobile and check Site Settings → should be Gutenberg
⏳ Create a new WP.com site on mobile and start a new post → should be Gutenberg and show the info popup. Result: it's Gutenberg but no popup.

⏳ Clean app, clean backend, start new post → should be Aztec
⏳ Clean app, "Aztec" in backend, start new post → should be Aztec
⏳ Clean app, "Gutenberg" in backend, start new post → should be Gutenberg
⏳ Clean  app, clean backend, open Gutenberg post → auto-enables Gutenberg
⏳ Clean  app, clean backend, open Gutenberg post (auto-enables GB), start new post → should be Gutenberg
⏳ Clean  app, clean backend, open Gutenberg post (auto-enables GB), open a GB post again→ Gutenberg but no popup
⏳ Clean app, clean backend, open Gutenberg post (auto-enables GB), new post in other site → should be Aztec
⏳ Clean app, clean backend, open Gutenberg post (auto-enables GB), open GB post in other site → auto-enables Gutenberg

⏳ Toggling the setting in a site doesn't toggle it in others

⏳ Migrate from v12.9 with GB ON, start new post → should be Gutenberg in all sites: `seems to be Aztec on current site and Gutenberg on other site`
⏳ Migrate from v12.9 with GB OFF and been auto-enabled, start new post → should be Aztec in all sites
⏳ Migrate from v12.9 with GB OFF and never auto-enabled, start new post → should be Aztec in all sites
⏳ Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (should auto-enable GB), change to other site and start new post → should be Aztec
⏳ Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (should auto-enable GB), change to other site and open GB post → auto-enables Gutenberg on the site

## Selfhosted sites

⏳ Self-hosted site added, start new post → should be Aztec
⏳ Self-hosted site added, open a GB post → auto-enables Gutenberg
⏳ Self-hosted site added, open a GB post (should auto-enable GB), start new post → should be Gutenberg
⏳ Self-hosted site added, open a GB post (should auto-enable GB), clean app data, re-add self-hosted site, start new post → should be Aztec
⏳ Self-hosted site added, open a GB post (should auto-enable GB), clean app data, re-add self-hosted site, open a GB post → auto-enables Gutenberg

⏳ Migrate a selfhosted from v12.9 with GB OFF and been auto-enabled, start new post → should be Aztec in all selfhosted sites
⏳ Migrate from v12.9 with GB OFF and never auto-enabled, start new post → should be Aztec in all selfhosted sites
⏳ Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (should auto-enable GB), change to other site and start new post → should be Aztec
⏳ Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (should auto-enable GB), change to other site and open GB post → auto-enables Gutenberg on the site

daniloercoli commented:
Tested with self-hosted sites!

✅ Self-hosted site added, start new post → should be Aztec
✅ Self-hosted site added, open a GB post → auto-enables Gutenberg
✅ Self-hosted site added, open a GB post (should auto-enable GB), start new post → should be Gutenberg
✅ Self-hosted site added, open a GB post (should auto-enable GB), clean app data, re-add self-hosted site, start new post → should be Aztec
✅ Self-hosted site added, open a GB post (should auto-enable GB), clean app data, re-add self-hosted site, open a GB post → auto-enables Gutenberg

✅ Migrate a selfhosted from v12.9 with GB OFF and been auto-enabled, start new post → should be Aztec in all selfhosted sites
✅ Migrate from v12.9 with GB OFF and never auto-enabled, start new post → should be Aztec in all selfhosted sites
✅ Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (should auto-enable GB), change to other site and start new post → should be Aztec
✅ Migrate from v12.9 with GB OFF and never auto-enabled, open a GB post (should auto-enable GB), change to other site and open GB post → auto-enables Gutenberg on the site

hypest commented:
The [WPiOS PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/12221) is also approved now so, merging these both.

daniloercoli commented:
Great team working on this PR!
